File: latte/src/main/java/gg/beemo/latte/broker/KafkaMessageHeaders.kt
package gg.beemo.latte.broker

import gg.beemo.latte.broker.IBrokerMessageHeaders.Companion.HEADER_CLIENT_ID
import gg.beemo.latte.broker.IBrokerMessageHeaders.Companion.HEADER_REQUEST_ID
import gg.beemo.latte.broker.IBrokerMessageHeaders.Companion.HEADER_SOURCE_CLUSTER
import gg.beemo.latte.broker.IBrokerMessageHeaders.Companion.HEADER_TARGET_CLUSTERS
import org.apache.kafka.common.header.Headers
import java.util.UUID

class KafkaMessageHeaders(
    override val clientId: String,
    override val sourceCluster: String,
    targetClusters: Set<String>?,
    requestId: String?,
) : IBrokerMessageHeaders {

    override val targetClusters: Set<String> = targetClusters ?: emptySet()
    override val requestId: String = requestId ?: UUID.randomUUID().toString()

    constructor(headers: Headers) : this(
        headers.getOrDefault(HEADER_CLIENT_ID, "default"),
        headers.getOrDefault(HEADER_SOURCE_CLUSTER, INVALID_CLUSTER_ID.toString()),
        headers.getOrDefault(HEADER_TARGET_CLUSTERS, "")
            .split(",")
            .filter { it.isNotEmpty() }
            .toSet(),
        headers.getOrDefault(HEADER_REQUEST_ID, ""),
    )

    fun applyTo(headers: Headers) {
        headers.add(HEADER_CLIENT_ID, clientId.toByteArray())
        headers.add(HEADER_TARGET_CLUSTERS, targetClusters.joinToString(",").toByteArray())
        headers.add(HEADER_REQUEST_ID, requestId.toByteArray())
        headers.add(HEADER_SOURCE_CLUSTER, sourceCluster.toByteArray())
    }

    companion object {

        const val INVALID_CLUSTER_ID = Integer.MIN_VALUE

    }

}

private fun Headers.getOrDefault(key: String, defaultValue: String): String {
    return String(lastHeader(key)?.value() ?: defaultValue.toByteArray())
}

File: water/src/broker/kafka/KafkaMessageHeaders.ts
import { randomUUID } from "node:crypto";
import type { IHeaders } from "kafkajs";
import type { IBrokerMessageHeaders } from "../BrokerMessageHeaders.js";
import { BROKER_HEADERS } from "../BrokerMessageHeaders.js";

export class KafkaMessageHeaders implements IBrokerMessageHeaders {
	public static readonly INVALID_CLUSTER_ID = `${-Math.pow(2, 32)}`;

	public constructor(
		public readonly clientId: string,
		public readonly sourceCluster: string,
		public readonly targetClusters: Set<string> = new Set(), // Empty = No target restriction
		public readonly requestId: string = randomUUID()
	) {}

	public serialize(): IHeaders {
		return {
			[BROKER_HEADERS.CLIENT_ID]: this.clientId,
			[BROKER_HEADERS.TARGET_CLUSTERS]: [...this.targetClusters].join(","),
			[BROKER_HEADERS.REQUEST_ID]: this.requestId,
			[BROKER_HEADERS.SOURCE_CLUSTER]: this.sourceCluster.toString(),
		};
	}

	public static from(headers: IHeaders): KafkaMessageHeaders {
		const transformed = KafkaMessageHeaders.headersBufferToString(headers);
		return new KafkaMessageHeaders(
			transformed[BROKER_HEADERS.CLIENT_ID] ?? "default",
			transformed[BROKER_HEADERS.SOURCE_CLUSTER] || KafkaMessageHeaders.INVALID_CLUSTER_ID,
			new Set((transformed[BROKER_HEADERS.TARGET_CLUSTERS] ?? "").split(",").filter(c => c)),
			transformed[BROKER_HEADERS.REQUEST_ID] ?? ""
		);
	}

	private static headersBufferToString(headers: IHeaders): Record<string, string> {
		return Object.fromEntries(
			Object.entries(headers)
				.filter(([_, v]) => v)
				.map(([k, v]) => [k, v!.toString()])
		);
	}
}
